### PR TITLE
Add tracepoint for publish/subscribe serialized message

### PR DESCRIPTION
**Pull request by h-suzuki-isp**
Please, refer to https://github.com/ros2/rclcpp/pull/2448 .
I have added the following trace points.

- `__rmw_publish_serialized_message()`
  - tracepoint : rmw_publish
- `_take_serialized_message()`
  - tracepoint : rmw_take

### CODE REVIEW

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 1]
```diff
@@ -3553,12 +3568,24 @@ static rmw_ret_t rmw_take_ser_int(
...
       }
     }
     ddsi_serdata_unref(d);
   }
   *taken = false;
+  TRACETOOLS_TRACEPOINT(
+    rmw_take,
+    static_cast<const void *>(subscription),
+    static_cast<const void *>(serialized_message),
+    (message_info ? message_info->source_timestamp : 0LL),
```
Comment (eboasson) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp:3587]: If nothing was taken, `message_info->source_timestamp` is meaningless.

Wouldn't:
```suggestion
    0LL,
```
make more sense?

That is, of course, under the assumption that you really want the tracepoint to fire when successfully taking nothing, rather than firing only when actually taking something.

Comment (h-suzuki-isp) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp:3587]: Thank you for your review, I required.

**Review by christophebedard — APPROVED** [review]
This looks good to me. @eboasson or @mjcarroll do you want to take a(nother) look?

**Comment by h-suzuki-isp** [comment]
@eboasson @mjcarroll (Cc: @christophebedard )
I'd like to get this PR to Jazzy on time.
I would appreciate it if you could review it.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 2]
```diff
@@ -2040,7 +2044,7 @@ extern "C" rmw_ret_t rmw_publish_serialized_message(
   }
 #endif
 
-  const bool ok = (dds_writecdr(pub->enth, d) >= 0);
+  const bool ok = (dds_forwardcdr(pub->enth, d) >= 0);
```
Comment (mjcarroll) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp:2047]: Just verifying this: we are switching from `writecdr` to `forwardcdr` because we are setting the timestamp above?  Is there any potential fallout from this change?

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 3]
```diff
@@ -3506,7 +3510,6 @@ static rmw_ret_t rmw_take_ser_int(
       if (message_info) {
         message_info_from_sample_info(info, message_info);
       }
-
```
Comment (mjcarroll) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp:3509]: Nit: unnecessary whitespace change

**Review by mjcarroll — CHANGES_REQUESTED** [review]
Mostly good, one nit and I would like clarification on the impact of the API change.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 2 above)
Comment (christophebedard) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp:2047]: Apparently not, the same thing was done in #454.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 3 above)
Comment (h-suzuki-isp) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Fixed in https://github.com/ros2/rmw_cyclonedds/pull/485/commits/2b33c0bea1e4045d5f8c86ee2184b16f297c6b99.

**Comment by christophebedard** [comment]
I think we can probably skip CI since the only new change is a new blank line.

**Review by mjcarroll — APPROVED** [review]
